### PR TITLE
Fix defining and appending multi-controlled gates with 0 control qubits.

### LINKED ISSUE #15664: mcrz with empty controls causes memory allocation failure (Core Dump)
### Environment

- **Qiskit Version**: 2.3.0
- **OS**: Linux
- **Python Version**: 3.10

### What is happening?

Calling `QuantumCircuit.mcrz` with an empty list of control qubits (`[]`) causes the Python process to crash with a "memory allocation of 68719476720 bytes failed" error (Core Dump).

### How can we reproduce the issue?

Run the following script:
```python
from qiskit import QuantumCircuit
qc = QuantumCircuit(1)
# This triggers a memory allocation of ~64GB+ and crashes
qc.mcrz(0.1, [], 0)

### What should happen?

The code should run successfully (equivalent to a standard rz gate) or raise a descriptive error. It should not crash the interpreter.

### Any suggestions?

_No response_

### LINKED ISSUE #15666: `RZGate.control(num_ctrl_qubits=0)` causes unbounded memory allocation and crash
### Environment

- **Qiskit version**: 2.3.0
- **Python version**: 3.11.14
- **Operating system**: Ubuntu 22.04

### What is happening?

Calling `RZGate.control` with `num_ctrl_qubits=0` leads to unbounded memory allocation. The process continuously consumes memory and eventually crashes with a `core dumped` error. This boundary case is not validated and results in catastrophic failure rather than a controlled error.

### How can we reproduce the issue?

```python
from qiskit.circuit.library.standard_gates.rz import RZGate
rz_gate_no_control = RZGate(0.1).control(num_ctrl_qubits=0)
```
Executing this code causes memory usage to grow without bound and eventually terminates the process.

### What should happen?

The method should either raise a clear `ValueError` for `num_ctrl_qubits=0` or handle this case explicitly, but must not allocate unbounded memory or crash.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by alexanderivrii**
### Summary

Fixes #15664 and #15666. 

### Details and comments

Previously, calling `gate.control(num_ctrl_qubits=0)` for various gates could sometimes raise an error, sometimes cause a core dump, and sometimes work without problems. Similarly, calling `QuantumCircuit` appending method for multi-controlled gates with no controls could sometimes work correctly but produce inefficient circuits, sometimes raise an error, and sometimes cause a memory overflow. 

While controlling gates with 0 controlled qubits is somewhat of an edge case, it could be nice to support this. 

Now, `gate.control(num_ctrl_qubits)=0` returns a copy of the gate and `qc.mcrz` and friends specially handle the case when there are no control qubits.

I have also improved error messages when a negative number of control qubits is specified. 

In addition, testing discovered a very old bug in the definition of the MCPhase gate with zero controls: the definition consisted of two phase gates instead of one.

### CODE REVIEW

File: qiskit/circuit/quantumcircuit.py [hunk 1]
```diff
@@ -5477,7 +5483,11 @@ def mcrz(
         self._check_dups(all_qubits)
 
         n_c = len(control_qubits)
-        if n_c == 1:
+
+        if n_c == 0:
+            self.append(RZGate(lam), [target_qubit])
```
Comment (ShellyGarion) [inline qiskit/circuit/quantumcircuit.py]: why not `self.rz(lam, target_qubit)` ?

File: qiskit/circuit/library/standard_gates/p.py [hunk 2]
```diff
@@ -360,7 +360,7 @@ def _define(self):
 
         if self.num_ctrl_qubits == 0:
             qc.p(self.params[0], 0)
-        if self.num_ctrl_qubits == 1:
+        elif self.num_ctrl_qubits == 1:
```
Comment (ShellyGarion) [inline qiskit/circuit/library/standard_gates/p.py:363]: should it raise an error for negative number of controls? or is it too much?

**Review by ShellyGarion — COMMENTED** [review]
@alexanderivrii - thanks for fixing it bug and making a consistent behaviour of all multi-controlled gates. I have some minor comments.

Would you like to add a test that asserts that calling them with negative number of qubits raises an error? or is it too much?

File: qiskit/circuit/quantumcircuit.py (hunk 1 above)
Comment (alexanderivrii) [reply qiskit/circuit/quantumcircuit.py]: Sure, done in [5dcfa8e](https://github.com/Qiskit/qiskit/pull/15673/commits/5dcfa8eb649a4428791a5db154ed4f5dffda3675).

File: qiskit/circuit/library/standard_gates/p.py (hunk 2 above)
Comment (alexanderivrii) [reply qiskit/circuit/library/standard_gates/p.py:363]: When we call `_define`, we can assume that the gate is valid, so there is no need to raise an error.

And if we try to instantiate an `MCPhaseGate` with a negative number of control qubits, e.g.
```python
gate = MCPhaseGate(0.1, num_ctrl_qubits=-1)
```
we will get a nice error message
```
qiskit.circuit.exceptions.CircuitError: 'The number of control qubits must be non-negative.'
```

File: test/python/circuit/test_controlled_gate.py [hunk 3]
```diff
@@ -650,6 +650,36 @@ def test_multi_controlled_rotation_gate_matrices(
             with self.subTest(msg=f"control state = {ctrl_state}"):
                 self.assertTrue(matrix_equal(simulated, expected))
 
+    def test_append_mcrx_with_0_controls(self):
```
Comment (ShellyGarion) [inline test/python/circuit/test_controlled_gate.py:653]: perhaps we should unify all these tests into a single test?

**Comment by alexanderivrii** [comment]
> Would you like to add a test that asserts that calling them with negative number of qubits raises an error? or is it too much?

Interesting. We already have a test called `"test_improper_num_ctrl_qubits"` that checks the validity of the `num_ctrl_qubits` argument for the `ControlledGate` class, but it was broken: it automatically returned `CircuitError` since the argument `base_gate` was not specified. In other words, it was expected that this test should return `CircuitError` but it did so for a different reason. I have fixed this test (and note that it includes checking that passing a negative integer number of control qubits is invalid). I have also added a similar test for the `gate.control` method.

File: test/python/circuit/test_controlled_gate.py (hunk 3 above)
Comment (alexanderivrii) [reply test/python/circuit/test_controlled_gate.py:653]: It is a nice idea, but I can't think of a good way to do so. For example, I _could_ replace `qc.mcrx` by `getattr(qc, "mcrx")` but the intent of the test becomes less clear. I could also create a single test with multiple subtests if you think it's more clear.

**Review by ShellyGarion — APPROVED** [review]
